### PR TITLE
Make GC occasionally refresh its discovery information

### PR DESCRIPTION
**Pull request by lavalamp**
ref #44507 - this should unblock AA and TPR even though it's not a full fix.

### CODE REVIEW

File: pkg/controller/garbagecollector/memcachediscovery/client.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+func (d *Client) ServerGroups() (*metav1.APIGroupList, error) {
+	if g, ok := func() (*metav1.APIGroupList, bool) {
+		d.lock.RLock()
+		defer d.lock.RUnlock()
+		return d.groupList, d.groupList != nil
+	}(); ok {
+		return g, nil
+	}
+
+	liveGroups, err := d.delegate.ServerGroups()
```
Comment (deads2k) [inline pkg/controller/garbagecollector/memcachediscovery/client.go]: We've had issues caching empty returns.  If the list is empty, don't cache it.  Same fo rresources.

File: pkg/controller/garbagecollector/memcachediscovery/client.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+func (d *Client) SwaggerSchema(version schema.GroupVersion) (*swagger.ApiDeclaration, error) {
+	return d.delegate.SwaggerSchema(version)
+}
+
+func (d *Client) OpenAPISchema() (*spec.Swagger, error) {
+	return d.delegate.OpenAPISchema()
+}
+
+func (d *Client) Fresh() bool {
+	return true
```
Comment (deads2k) [inline pkg/controller/garbagecollector/memcachediscovery/client.go]: this looks off.  Why are you always fresh?  Deserves a comment if its right.

File: pkg/controller/garbagecollector/memcachediscovery/client.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+		defer d.lock.RUnlock()
+		return d.refresh()
+	}()
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	d.groupToServerResources, d.groupList = rl, gl
+}
+
+// d must be RLocked or Locked.
```
Comment (deads2k) [inline pkg/controller/garbagecollector/memcachediscovery/client.go]: why?  Just for the cached return on error?  I'd rather see that handled by outter logic that handles the error.

File: pkg/controller/garbagecollector/memcachediscovery/client.go (hunk 2 above)
Comment (deads2k) [inline pkg/controller/garbagecollector/memcachediscovery/client.go]: Alright, I think I see.  Setting this means that you're never being invalidated by higher logic.  Definitely needs a comment explaining it and how its to be used.

**Comment by deads2k** [comment]
I think that this needs at least one e2e test that creates a TPR and proves that GC collects it.

File: pkg/controller/garbagecollector/memcachediscovery/client.go (hunk 2 above)
Comment (lavalamp) [inline pkg/controller/garbagecollector/memcachediscovery/client.go]: Honestly the interface didn't document what this is supposed to do...

File: pkg/controller/garbagecollector/memcachediscovery/client.go (hunk 1 above)
Comment (lavalamp) [inline pkg/controller/garbagecollector/memcachediscovery/client.go]: What causes empty lists?

This is going to be updated every 30s so it should be ok if it forgets stuff for 30s

**Comment by lavalamp** [comment]
thanks for feedback.

we've got to stop making unwieldy e2e tests. This should be testable as a unit or integration test in the GC, which I should add.

File: pkg/controller/garbagecollector/memcachediscovery/client.go (hunk 1 above)
Comment (deads2k) [inline pkg/controller/garbagecollector/memcachediscovery/client.go]: > What causes empty lists?
> 
> This is going to be updated every 30s so it should be ok if it forgets stuff for 30s

Server start seems to start listening for discovery before registering groups.  It would be pretty bad if GC didn't work for a 30 second window for core types on startup.  It ought not be a difficult check to add.

Comment (lavalamp) [inline pkg/controller/garbagecollector/memcachediscovery/client.go]: seems like the wrong fix, server could be in the middle of registering groups, too, and then we'd get some right and some wrong for 30s.

If an empty list is an error, then apiserver should probably not return 200s on empty lists?

File: pkg/controller/garbagecollector/memcachediscovery/client.go (hunk 2 above)
Comment (lavalamp) [inline pkg/controller/garbagecollector/memcachediscovery/client.go]: Reading the code that calls Fresh I still don't understand what it is supposed to do.

**Comment by ncdc** [comment]
@lavalamp I don't think this is sufficient - the graph builder currently only sets up its monitors to list/watch resources when the GC controller is first instantiated, but it doesn't refresh right now.

**Comment by lavalamp** [comment]
@ncdc good point, I will fix that in another PR.

File: pkg/controller/garbagecollector/memcachediscovery/client.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+func (d *Client) OpenAPISchema() (*spec.Swagger, error) {
+	return d.delegate.OpenAPISchema()
+}
+
+func (d *Client) Fresh() bool {
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	// Fresh is supposed to tell the caller whether or not to retry if the
```
Comment (caesarxuchao) [inline pkg/controller/garbagecollector/memcachediscovery/client.go]: I think this is a better definition of Fress(). Could you update the [comment](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/discovery/discovery_client.go#L58) in the interface as well?

The current comment in the interface describes a narrower case of implemented by `CachedDiscoveryClient`, could you move the comment to [there](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/cached_discovery.go#L244)?

**Comment by lavalamp** [comment]
@mikedanese any idea why bazel can't find the new import?

File: pkg/controller/garbagecollector/memcachediscovery/client.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+		}
+	}
+
+	d.groupToServerResources, d.groupList = rl, gl
+	d.cacheValid = true
+}
+
+// NewClient creates a new Client which caches discovery information in memory
+// and will stay up-to-date if Invalidate is called with regularity.
+func NewClient(delegate discovery.DiscoveryInterface) *Client {
```
Comment (krmayankk) [inline pkg/controller/garbagecollector/memcachediscovery/client.go]: @lavalamp what does DiscoveryInterface provide ?  Is this related to service discovery or api discovery ?

File: cmd/kube-controller-manager/app/core.go [hunk 6]
```diff
@@ -170,10 +173,28 @@ func startGarbageCollectorController(ctx ControllerContext) (bool, error) {
...
+	restMapper.Reset()
+	go func() {
+		t := time.NewTicker(30 * time.Second)
+		defer t.Stop()
+		for {
+			select {
+			case <-t.C:
+				restMapper.Reset()
+			case <-ctx.Stop:
+				return
```
Comment (deads2k) [inline cmd/kube-controller-manager/app/core.go]: stop your ticker.

Comment (deads2k) [inline cmd/kube-controller-manager/app/core.go]: > stop your ticker.

Oh, I see the defer

**Comment by mikedanese** [comment]
Looks like you figured it out.

**Comment by lavalamp** [comment]
@mikedanese yes, I did, sorry for the noise. (git commit -a doesn't actually add new files, duh...)

File: pkg/controller/garbagecollector/memcachediscovery/client.go (hunk 4 above)
Comment (lavalamp) [inline pkg/controller/garbagecollector/memcachediscovery/client.go]: The current comment makes no sense to me whatsoever, so ok.

File: pkg/controller/garbagecollector/memcachediscovery/client.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+var _ discovery.CachedDiscoveryInterface = &Client{}
+
+// ServerResourcesForGroupVersion returns the supported resources for a group and version.
+func (d *Client) ServerResourcesForGroupVersion(groupVersion string) (*metav1.APIResourceList, error) {
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	cachedVal, ok := d.groupToServerResources[groupVersion]
+	if !ok {
+		return nil, ErrCacheEmpty
```
Comment (sttts) [inline pkg/controller/garbagecollector/memcachediscovery/client.go]: looks more like a NotFound than ErrCacheEmpty.

File: cmd/kube-controller-manager/app/core.go [hunk 8]
```diff
@@ -170,29 +174,62 @@ func startGarbageCollectorController(ctx ControllerContext) (bool, error) {
...
-		return true, fmt.Errorf("Failed to parse resources from server: %v", err)
+	getDeletableResources := func() (map[schema.GroupVersionResource]struct{}, error) {
+		preferredResources, err := gcClientset.Discovery().ServerPreferredResources()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get supported resources from server: %v", err)
+		}
+		deletableResources := discovery.FilteredBy(discovery.SupportsAllVerbs{Verbs: []string{"delete"}}, preferredResources)
+		deletableGroupVersionResources, err := discovery.GroupVersionResources(deletableResources)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse resources from server: %v", err)
```
Comment (sttts) [inline cmd/kube-controller-manager/app/core.go]: nit: lower-case

**Comment by deads2k** [comment]
@lavalamp are you going to come back to this as a bugfix after freeze?

**Comment by lavalamp** [comment]
@deads2k I am not sure. My calendar is pretty booked over the next two weeks. :/ Let me see if Dawn will buy this is a bug fix.

**Comment by ash2k** [comment]
Please consider moving the client to client-go repo so that it is more reusable. See kubernetes/client-go/issues/200

**Comment by ironcladlou** [comment]
@lavalamp do you have any update on your plans for this PR? Is there anything I can do to help move it forward?

File: pkg/controller/garbagecollector/graph_builder.go [hunk 9]
```diff
@@ -92,6 +101,45 @@ type GraphBuilder struct {
...
+
+func (sm *stopMux) stopper() {
+	defer close(sm.outputStop)
+	select {
+	case <-sm.globalStop:
+	case <-sm.localStop:
+	}
+}
+
+type monitorSingle struct {
```
Comment (sttts) [inline pkg/controller/garbagecollector/graph_builder.go]:
```
// monitorSingle wraps a cache.Controller and adds another stop channel (in addition to that of Run(stopCh)). It will stop
// if either of them is closed. 
```

File: pkg/controller/garbagecollector/graph_builder.go [hunk 10]
```diff
@@ -170,37 +219,71 @@ func (gb *GraphBuilder) controllerFor(resource schema.GroupVersionResource, kind
...
+		// TODO: remove all resources of this type from the graph?
 	}
-	return nil
+
+	// NewAggregate returns nil if errs is 0-length
+	return utilerrors.NewAggregate(errs)
 }
 
 func (gb *GraphBuilder) HasSynced() bool {
 	for _, monitor := range gb.monitors {
```
Comment (sttts) [inline pkg/controller/garbagecollector/graph_builder.go]: this misses a lock. Or at the very least it should be documented that it assumes it being taken.

File: pkg/controller/garbagecollector/graph_builder.go [hunk 11]
```diff
@@ -170,37 +219,71 @@ func (gb *GraphBuilder) controllerFor(resource schema.GroupVersionResource, kind
...
 			return false
 		}
 	}
 	return true
 }
 
 func (gb *GraphBuilder) Run(stopCh <-chan struct{}) {
+	gb.monitorLock.Lock()
+	defer gb.monitorLock.Unlock()
+	gb.globalStop = stopCh
```
Comment (sttts) [inline pkg/controller/garbagecollector/graph_builder.go]: Something is fishy: `NewGarbageCollector` calls `syncMonitors` which depends on `globalStop`.

File: pkg/controller/garbagecollector/graph_builder.go [hunk 12]
```diff
@@ -92,6 +101,45 @@ type GraphBuilder struct {
 	absentOwnerCache *UIDCache
 }
 
+type stopMux struct {
```
Comment (sttts) [inline pkg/controller/garbagecollector/graph_builder.go]: this `stopMux` thing does not make the code easier to understand IMO.  Just inline the mux logic into `monitorSingle.Run`.

It might even be much simpler to have one Go routine waiting on the global lock and shutting down all monitors in a loop. Any special reason that you want to have this logic per monitor?

File: pkg/controller/garbagecollector/metaonly/metaonly.go [hunk 13]
```diff
@@ -43,7 +43,39 @@ func gvkToMetadataOnlyObject(gvk schema.GroupVersionKind) runtime.Object {
 	}
 }
 
-func NewMetadataCodecFactory() serializer.CodecFactory {
+type MetadataCodecFactory struct {
```
Comment (sttts) [inline pkg/controller/garbagecollector/metaonly/metaonly.go]: what is this thing doing?

File: pkg/controller/garbagecollector/setup/setup.go [hunk 14]
```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (sttts) [inline pkg/controller/garbagecollector/setup/setup.go]: 2017

**Comment by ironcladlou** [comment]
This PR was superseded by https://github.com/kubernetes/kubernetes/pull/47665 and can be closed.
